README.md
# Sybase

#### 介绍
Sybase 数据存储分析和修复

#### 目录结构
目录结构说明

 代码         | 调用函数 | 说明
------------|------|---
 unload_dat |      | 解析dat文件
 test       |      | 测试代码和数据


#### 安装教程

python 3.6+

pip install -r requirements.txt

#### 使用说明




CHANGELOG
v0.1.0  2022.4.23
上传到github repo




unload_dat/test1.py
# -*- coding: utf-8 -*-
import os.path, struct, sqlite3
import table_struct, pymssql, page

s = struct.unpack


# 测试 表的页面 链接
def init_link(f, tab_data_07):
    tab_data_22 = 0
    for i in range(0, len(tab_data_07)):
        for ii in range(0, tab_data_07[i].rec_sum):
            if tab_data_07[i].record[ii].col_data1[2] == 281474978938880 and tab_data_07[i].record[ii].col_data1[
                1] == 1:  # 22 表
                # ownerid：281474978938880 （22表） 327680 （05表）458752（07表） 281474979397632（29表） 281474977562624（0D表）281474979987456（1D表）
                pgfirst = tab_data_07[i].record[ii].col_data2[5]
                print(pgfirst)
                tab_data_22 = page_link1(f, pgfirst)
                print('afdff')


def page_link1(f, pgfirst):  # 文件，要解析的页链的起始页面号， 测试 链接
    page_size = 8192
    f_size = os.path.getsize(f.name)  # 获取文件大小，只能处理文件,不能磁盘
    f_pg_size = f_size // 8192
    next_page = pgfirst
    while (next_page != 0 and next_page < f_pg_size):  # 下一页不溢出
        pos1 = (next_page) * page_size
        f.seek(pos1)  # 要判断返回值，是否成功
        data = f.read(page_size)
        data1 = data[0:page_size]
        data3 = s('<I', data1[24:28])  # obj_id
        print('page_no: %d, obj_id: %d ===========' % (next_page, data3[0]))
        #  page1 = page.record(data1,table)  # 解析页面记录.  ***************
        data2 = s('<I', data1[16:20])
        next_page = data2[0]  # 下一页的页号


# 扫描所有的页面，解析obj_id符合的页面
def page_scan(f, tab):
    page_size = 8192;
    page_sum = 0
    f_size = os.path.getsize(f.name)  # 获取文件大小，只能处理文件,不能磁盘
    f_pg_size = f_size // 8192
    tables = [];
    tab_data = []
    tables.append(tab)
    for i in range(0, f_pg_size):  # 下一页不溢出
        pos1 = i * page_size
        f.seek(pos1)  # 要判断返回值，是否成功
        data = f.read(page_size)
        data1 = data[0:page_size]
        data2 = s('<H', data1[6:8])  # ind_id，auid_ind_id
        data3 = s('<I', data1[24:28])  # obj_id，auid_obj_id
        data4 = s('<I', data1[16:20])  # page_no
        for tab in tables:
            if data3[0] == tab.auid_obj_id and data2[0] == 256:
                page_sum += 1
                #   tab.pgfirst = page_sum
                page1 = page.record_2(f, data1, tab)  # 解析页面记录.  ***************
                tab_data.append(page1)
        del data, data1
    print('i:%d, page_no:%d, obj_id:%d, ind_id:%d, page_sum:%d ===========\r' % (
        i, data4[0], data3[0], data2[0], page_sum))
    return tab_data


def init_tab(db):
    # 读取sqlite数据库表，获得配置信息 和 表结构信息
    # db = './db_info.db'           # 数据库文件名称
    conn = sqlite3.connect(db)  # 打开数据库
    cursor = conn.cursor()
    cursor.execute("select * from tab_info ")
    values_tab = cursor.fetchall()  # 返回数据为二维数组
    tables = []  #
    for i in range(len(values_tab)):  # 初始化表结构
        table1 = table_struct.st_table()
        table1.tab_obj_id = values_tab[i][1]
        table1.tab_name = values_tab[i][3]
        table1.col_sum = values_tab[i][4]
        table1.nullable_sum = values_tab[i][5]
        cursor.execute(
            "select c.* from tab_info t,col_info c where c.tab_obj_id = t.tab_obj_id and t.tab_obj_id=%d order by c.col_id" % table1.tab_obj_id)
        values_col = cursor.fetchall()
        if len(values_col) == 0:
            table1.col_sum = 0
        var_len_sum = 0  # 变长列数
        for ii in range(0, table1.col_sum):
            column1 = table_struct.st_column()
            column1.tab_obj_id = values_col[ii][1]
            column1.col_id = values_col[ii][2]
            column1.col_name = values_col[ii][3]
            column1.col_x_type = values_col[ii][4]
            column1.col_u_type = values_col[ii][5]
            column1.col_len = values_col[ii][6]
            column1.varlen_is = values_col[ii][12]
            table1.col.append(column1)
            if column1.varlen_is == 0:
                table1.col_1.append(column1)
            elif column1.varlen_is == 1:
                var_len_sum = var_len_sum + 1
        table1.var_len_sum = var_len_sum
        for ii in range(0, table1.col_sum):
            if table1.col[ii].varlen_is == 1:
                column1 = table1.col[ii]
                table1.col_1.append(column1)
        tables.append(table1)
    cursor.close()
    conn.close()
    return tables

# f = open (r'h:\aaaa.mdf','rb')
# db = r'.\tables.db'
# page_scan(f,db)

# conn = pymssql.connect(host="127.0.0.1\MSSQL2005",user="sa",password="bysjhf808.",database="master")
# cur=conn.cursor()
# cur.execute('select top 1 * from sysobjects;')
# print(cur.fetchall())
# data=cur.fetchall()


unload_dat/data_type.py
# -*- coding: utf-8 -*-
# 数据类型解析，持续完善
import struct, time, decimal

s = struct.unpack  # B H I


def data_type(data, col_1):
    if len(data) == 0:
        return 0
    # 整数 int,bigint,smallint,tinyint ...
    if col_1.col_x_type in (1, 48, 52, 56, 127):
        if col_1.col_len == 1:
            a = s("<B", data)
        elif col_1.col_len == 2:
            a = s("<h", data)
        elif col_1.col_len == 4:
            a = s("<i", data)
        elif col_1.col_len == 8:
            a = s("<q", data)
        data_out = a[0]
        return data_out

    # 字符 char,varchar ...
    elif col_1.col_x_type in (2, 175, 167):
        data_out = ''
        try:
            data_out = str(data, encoding="gbk").rstrip()  # ascii,gbk,gb2312   .strip()
        except UnicodeDecodeError:
            data_out = ''
            return data_out
        if col_1.col_len == 2 and data == b'\x00\x00':
            data_out = ''
        elif col_1.col_len == 1 and data == b'\x00':
            data_out = ''
        return data_out

    # 字符 nchar, nvarchar  ...
    elif col_1.col_x_type in (3, 231, 239):
        try:
            data_out = str(data, encoding="utf-16").rstrip()  # utf-8 ,utf-16 去字符串尾部的空格
        except UnicodeDecodeError:
            data_out = 'nvarchar溢出'
        return data_out
    elif col_1.col_x_type == 99:
        data_out = 'ntext'
        return data_out

    # 日期时间 datetime ,date ,time, smalldatetime, timestamp ...
    elif col_1.col_x_type == 58:  # smalldatetime
        a = s("<2H", data)
        date1 = a[1];
        time1 = a[0]
        time2 = date1 * 24 * 60 * 60 - 2209017600 + time1 * 60
        data_out = str(time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time2)))
        return data_out
    elif col_1.col_x_type == 40:  # date
        a = s("<HB", data)
        date = (a[1] * 65536 + a[0] - 693595) * 24 * 60 * 60 - 2209017600
        try:
            data_out = str(time.strftime('%Y-%m-%d', time.localtime(date)))
        except OSError:
            data_out = ''
        return data_out
    elif col_1.col_x_type == 41:  # time
        if len(data) == 5:
            a = s("<IB", data)
            time1 = (a[1] * 4294967296 + a[0]) / 10 / 1000 / 1000
            data_out = str(time.strftime('%H:%M:%S.', time.localtime(time1))) + ('%.7f' % (time1))[-7:]
        else:
            data_out = ''
        return data_out
    elif col_1.col_x_type == 61:  # datetime
        a = s("<Ii", data)
        date1 = a[1];
        time1 = a[0] / 300
        time2 = date1 * 24 * 60 * 60 - 2209017600 + time1
        try:
            data_out = str(time.strftime('%Y-%m-%d %H:%M:%S.', time.localtime(time2))) + ('%.3f' % (time1))[-3:]
        except OSError:
            data_out = ''
        #    print('时间溢出...')
        return data_out
    elif col_1.col_x_type in (189, 42):  # timestamp,datetime2,, 没有解析
        data_out = ''
        return data_out

    # numeric,decimal,money
    elif col_1.col_x_type in (108, 106):  # numeric,decimal
        data_out = 0
        if col_1.col_len == 9:
            data1 = s('<Q', data[1:9])
            prec = col_1.prec
            scale = col_1.scale
            data_1 = data1[0]
            data_out = decimal.Decimal(data_1 / decimal.Decimal(str(10 ** scale)))
        elif col_1.col_len == 5:
            data1 = s('<I', data[1:5])
            prec = col_1.prec
            scale = col_1.scale
            data_1 = data1[0]
            data_out = decimal.Decimal(data_1 / decimal.Decimal(str(10 ** scale)))
        elif col_1.col_len == 13:
            data1 = s('<Q', data[1:9])
            data2 = s('<I', data[9:13])
            prec = col_1.prec
            scale = col_1.scale
            data_1 = data2[0] * (256 ** 8) + data1[0]
            data_out = decimal.Decimal(data_1 / decimal.Decimal(str(10 ** scale)))
        elif col_1.col_len == 17:
            data1 = s('<Q', data[1:9])
            data2 = s('<Q', data[9:17])
            prec = col_1.prec
            scale = col_1.scale
            #    aa = '%%%d.%df'%(prec,scale)
            data_1 = data2[0] * (256 ** 8) + data1[0]
            #    print('%d %d,%d,%s'%(data2[0],data1[0],data_1,decimal.Decimal(data_1/decimal.Decimal(str(10**scale)))))
            data_out = decimal.Decimal(data_1 / decimal.Decimal(str(10 ** scale)))  # .strip()
        return data_out

    elif col_1.col_x_type == 62:  # fload
        if col_1.col_len == 8:
            data1 = s('<d', data[0:8])
            data_out = data1[0]
        return data_out
    elif col_1.col_x_type == 59:  # real
        if col_1.col_len == 4:
            data1 = s('<f', data[0:4])
            data_out = data1[0]
        return data_out
    elif col_1.col_x_type == 60:  # money
        if col_1.col_len == 8:
            try:
                data1 = s('<q', data[0:8])
                data_out = float(data1[0] / 10000)
            except struct.error:
                data_out = 0
            data_out = '%.4f' % (data_out)
        return data_out
    elif col_1.col_x_type == 122:  # smallmoney
        if col_1.col_len == 4:
            data1 = s('<i', data[0:4])
            data_out = float(data1[0] / 10000)
            data_out = '%.4f' % (data_out)
        return data_out

    # page_file,  binary
    elif col_1.col_x_type in (5, 173):
        if col_1.col_len == 6:
            data1 = s('<IH', data)
            return data1[0]
        else:
            return 0

    elif col_1.col_x_type in (-1, 98):
        try:
            data_out = str(data, encoding="gbk").rstrip()  # utf-8 ,utf-16 去字符串尾部的空格
        except UnicodeDecodeError:
            data_out = 'sql_variant溢出'
        return data_out

    # 其他 uniqueidentifier,geography,sql_variant,xml
    elif col_1.col_x_type in (36, 130, 241):
        return 0
    else:
        return 0


unload_dat/unload_dat.py
# -*- coding: utf-8 -*-
# 读取数据库文件进行解析,, 主函数 模块. 能解析 sybase12.x,15.x

import struct, time
import init

s = struct.unpack


class Unload_DB():
    def __init__(self):  # 解析函数
        super(Unload_DB, self).__init__()

    # 文件信息
    def file_init(self, fn):
        self.files = []
        for f_name in fn:  # 每个文件头信息
            f = open(f_name, 'rb')
            data = f.read(8192)
            file_info = init.file_blk_0(data)
            f.seek(8192 * 9)
            data = f.read(8192)
            version1, version2, first_07, db_name = init.file_blk_9(data)
            file_info.f_name = f_name
            file_info.file = f
            file_info.version1 = version1
            file_info.version2 = version2
            file_info.boot = first_07
            file_info.db_name = db_name
            self.files.append(file_info)
        return self.files

    def unload_db(self, file_infos, db):
        print("Datetime: " + time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time())))  # current time
        first_07 = file_infos[0].boot
        f = file_infos[0].file
        version = file_infos[0].version1
        tab_data_07 = init.init_07(file_infos[0], first_07)
        print("tab_data_07/02 over ....")
        if version[3:] != '0':
            print("tab_data_29 in ....")
            tab_data_29 = init.init_29(f, tab_data_07)
            print("tab_data_29 over ....")
            tab_data_22, tab_data_05 = init.init_22_05(f, tab_data_07, tab_data_29)
            print("tab_data_22,tab_data_05 over ....")
            tab_all, view_all, program_all, trigger_all, default_all = init.init_all_2008(f, tab_data_22, tab_data_29,
                                                                                          tab_data_05,
                                                                                          tab_data_07)  # 获得所有表的结构
            print("unload all over")
        elif version[3:] == '0':
            tab_data_2 = tab_data_07
            tab_data_1, tab_data_3 = init.init_1_3(f, tab_data_2)
            print("init_1_3 over ....")
            tab_all, view_all, program_all, trigger_all, default_all = init.init_all_2000(f, tab_data_1, tab_data_2,
                                                                                          tab_data_3)
            print("init_all_2000 over ....")

        return tab_all, view_all, program_all, trigger_all, default_all

    def unload_tab(self, file_infos, tab, db):
        f = file_infos[0].file
        tab_data = init.unload_tab(f, tab)
        return tab_data

    # save table
    def save_tab(self, tab_data, tab, db):
        if tab.tab_name != 'sysowners':
            init.save_tab(tab_data, tab, db)


unload_dat/table_struct.py
# -*- coding: utf-8 -*-
# 定义 表结构  # 重要
# create table db_info(db_id,db_version,os,tab_sum,create)
# create table tab_inf(db_id,tab_version,tab_type,tab_obj_id,tab_name,col_sum,nullable_sum,created,modified)
# create table col_inf(tab_obj_id,col_id,col_name,col_x_type,col_u_type,col_len,prec,scale,collationid,seed,nullable_is,pkey_is,var_len_is,def_data)

class File_Info():
    def __init__(self):
        self.f_name = ''  # 文件路径
        self.f_no = 0  # 文件号
        self.file = 0
        self.f_type = ''  # 文件类型
        self.blk_sum = 0  # 文件的块/页 总数
        self.db_name = 0  # 数据库名
        self.gp_id = 0  # 文件组ID
        self.gp_name = ''  # 文件组名
        self.version1 = 0  # 版本
        self.version2 = 0  # 版本
        self.boot = 0  # 启动页


class st_db:  # 数据库
    def __init__(self):
        self.db_id = 0  # 数据库 ID
        self.db_name = ''  # 数据库名
        self.db_version = ''  # 数据库版本  # mssql_2008, mysql_5.1 等
        self.page_size = 8192  # 页面大小,固定为8192字节
        self.endian = '<'  # 大小端. 0小端,1大端
        self.os = ''  # OS信息
        self.tab_sum = 0  # 表的数量
        self.tab = []  # 表


class st_table:  # 表
    def __init__(self):
        self.user_id = 0  # 用户 ID
        self.tab_obj_id = 0  # 表的 object_id
        self.auid_obj_id = 0  # 表的 obj_id
        self.auid_ind_id = 0  # 表的 ind_id
        self.tab_name = ''  # 表名  #有的表只有object_id没有表名
        self.col_sum = 0  # 表的总列数
        self.nullable_sum = 0  # 可为空的列数
        self.var_len_sum = 0  # 变长列的列数
        self.col = []  # 列信息
        #   self.created = ''         # 表的创建时间
        self.pgfirst = 0  # 表数据起始页号
        self.pgfirstiam = 0  # 表数据IAM起始页号
        self.sql = ''


class st_view:  # view
    def __init__(self):
        self.view_id = 0  # view ID
        self.view_name = ''  # view 名称
        self.user_id = 0


class st_program:  # program
    def __init__(self):
        self.program_id = 0  # program ID
        self.program_name = ''  # program 名称
        self.user_id = 0


class st_column:  # 列
    def __init__(self):
        self.tab_obj_id = 0  # 表的object_id
        self.col_id = 0  # 列id
        self.col_name = ''  # 列名
        self.col_x_type = 0  # 列的数据类型(系统)
        self.col_u_type = 0  # 列的数据类型(用户)
        self.col_type = ''  # 列的数据类型名称
        self.col_len = 0  # 列长度
        self.prec = 0  # 精度
        self.scale = 0  # 小数位数
        self.collationid = 0  # 排序规则编号
        self.seed = 0  # 自增种子.  标识种子+标识增量
        self.def_data = 'NULL'  # 默认值
        self.nullable_is = ''  # 是否可为空
        self.varlen_is = 0  # 是否是变长
        self.pkey_is = 0  # 是否是主键


class st_page:  # 页面
    def __init__(self):
        self.check = 0  # 校验和
        self.page_no = 0  # 页面编号，从0编号
        self.file_no = 1  # 文件号， 从1编号
        self.page_prev = 0  # 上一页
        self.page_next = 0  # 下一页
        self.page_type = 0  # 页面类型：数据页，IAM页，索引页
        self.obj_id = 0  # obj_id
        self.ind_id = 0  # ind_id
        self.rec_sum = 0  # 页面中记录数量,即slot数量
        self.page_level = 0  # 页面level
        self.flagBits = 0  # 页面校验标志
        self.page_slot = []  # 页尾slot
        self.record = []  # 页面中的记录


class st_record:  # 数据记录
    def __init__(self):
        self.null_map = []
        self.col_sum = 0
        self.var_col_sum = 0
        self.var_col_over = []  # col溢出
        self.var_col_off = []  # 边长列结束偏移列表
        self.col_data1 = []  # 记录的解析后列数据(表结构中的顺序)


unload_dat/page.py
# -*- coding: utf-8 -*-
# page ,  record
import struct
import table_struct, data_type

s = struct.unpack  # B,H,I


def page_handle(in_data):  # 处理残缺页校验的页数据,恢复校验前的
    check_value = s("<I", in_data[60:64])
    value = []
    for i in range(0, 16):
        value.append((check_value[0] >> (i * 2)) & 0x00000003)
    new_data = bytearray(8192)
    for i in range(0, 16):
        if i == 0:
            new_data[i * 512:i * 512 + 512] = in_data[i * 512:i * 512 + 512]
        else:
            magic = struct.unpack("B", in_data[i * 512 + 511:i * 512 + 512])
            if value[0] == (magic[0] & 0x03):
                new_data[i * 512:i * 512 + 511] = in_data[i * 512:i * 512 + 511]
                new_data[i * 512 + 511:i * 512 + 512] = struct.pack("B", (magic[0] & 0xFC) + value[i])
            else:
                print("Page Err:%d Sector Check Failure!" % i)  # 此处校验位不对
                new_data[i * 512:i * 512 + 511] = in_data[i * 512:i * 512 + 511]
                new_data[i * 512 + 511:i * 512 + 512] = struct.pack("B", (magic[0] & 0xFC) + value[i])
    return new_data


def page_init(in_data):  # 初始化页面信息
    page_size = 8192
    chk_type = s("B", in_data[5:6])  # 报这个错的，一般是 此页面数据没找到
    data = in_data
    if chk_type[0] % 4 == 1:  # 处理残缺页校验的页数据
        data = page_handle(in_data)
    page1 = table_struct.st_page()
    data1 = s("<4B2HIHHIHHI2HIH", data[0:38])
    page1.page_no = data1[15]
    page1.page_prev = data1[6]
    page1.page_next = data1[9]
    page1.page_type = data1[1]
    page1.obj_id = data1[12]
    page1.ind_id = data1[5]
    page1.page_level = data1[3]
    page1.rec_sum = data1[11]
    for i in range(0, page1.rec_sum):
        slot1 = s("<H", data[page_size - i * 2 - 2:page_size - i * 2])  # slot_off_list
        page1.page_slot.append(slot1[0])  # 一般每个页面里slot数量在 1-100.
    return page1, data


# 正常记录解析， bits, 行溢出 还没加(重构前有)
def record(f, in_data, table):  # 解析 compact 记录  in_data(输入页面数据),table(表结构)
    page1, in_data = page_init(in_data)  # 初始化页面
    for i in range(page1.rec_sum):  # 页中的所有记录
        record1 = table_struct.st_record()
        off_set = page1.page_slot[i]  # 记录偏移
        try:
            header1 = s("<2h", in_data[off_set:off_set + 4])  # 记录头
        except struct.error:
            continue
        header_a = header1[0]  # 记录类型
        if header_a in (12, 28, 60, 167):  # 删除记录
            continue
        off1 = header1[1] + off_set  # 列信息的位置偏移
        try:
            col_sum = s('<H', in_data[off1:off1 + 2])  # 记录中的列数
        except struct.error:
            continue
        record1.col_sum = col_sum[0]  # 此记录的总列数
        null_map_len = (record1.col_sum + 7) // 8  # null位图的长度
        null_map_len1 = (table.col_sum + 7) // 8
        if null_map_len > 10:  # 80列
            continue
        frm = '<' + str(null_map_len) + 'B'
        null_map = s(frm, in_data[off1 + 2:off1 + 2 + null_map_len])  # null位图  +++++++++++++++++++++++++++++++++
        null_map1 = []
        for i1 in range(null_map_len1):
            for i2 in range(0, 8):
                if i1 * 8 + i2 < record1.col_sum:
                    a = null_map[i1] >> i2
                    col_is_null = a % 2  # 此列是否为 null，1=null
                elif i1 * 8 + i2 >= record1.col_sum:
                    col_is_null = 1
                null_map1.append(col_is_null)  # 记录中的位图是描述的原表结构的null情况
        record1.null_map = null_map1  # null位图  ++++++++++++++++++++++++++++++++

        var_col_sum = s('<H', in_data[off1 + 2 + null_map_len:off1 + 4 + null_map_len])  # 变长列数量
        if table.var_len_sum == 0 or header_a == 16:  # 无变长列
            var_col_sum = [0, ]
        record1.var_col_sum = var_col_sum[0]  # 变长列数量
        for ii in range(record1.var_col_sum):  # 变长列 偏移
            try:
                var_col_off1 = s('<H', in_data[off1 + 4 + null_map_len + 2 * ii:off1 + 6 + null_map_len + 2 * ii])
            except struct.error:
                continue
            var_col_off = var_col_off1[0]
            var_col_over = 0
            if var_col_off > 32768:  # 字段 溢出,标志
                var_col_off -= 32768
                var_col_over = 1
            record1.var_col_over.append(var_col_over)
            record1.var_col_off.append(var_col_off)

        # if page1.obj_id == 1 :                      # 测试用
        #     print('i:%d, page_no:%d,obj_id:%d,ind_id:%d,level:%d,rec_sum:%d, off_set:%d,col_sum:%d,null_map_len:%d,null_map_len1:%d'%\
        #     (i,page1.page_no,page1.obj_id,page1.ind_id,page1.page_level,page1.rec_sum,off_set,record1.col_sum,null_map_len,null_map_len1))

        # 解析记录的各列数据, 行溢出先略,  ======================================
        col_off = off_set + 4  # 各列的起始偏移
        var_list = 0
        bit_1 = 0;
        bit_2 = 0;
        data_out = 0
        var_off = off1 + null_map_len + 4 + record1.var_col_sum * 2
        for ii3 in range(len(table.col)):
            len1 = table.col[ii3].col_len  # 列数据类型长度
            col_is_null = record1.null_map[ii3]  # 列是否为空
            if col_is_null == 0:  # 如果列 不为空
                if table.col[ii3].varlen_is == 0:  # 如果列 是定长
                    if table.col[ii3].col_x_type in (3, 239):  # nchar, ntext
                        len1 = len1 * 1
                    col_data_1 = in_data[col_off:col_off + len1]
                    col_off += len1
                elif table.col[ii3].varlen_is == 1:  # 是变长
                    if var_list == 0:  # 第一个变长
                        try:
                            len1 = record1.var_col_off[var_list] - header1[1] - (
                                    null_map_len + 4 + record1.var_col_sum * 2)
                        except IndexError:
                            continue
                    elif var_list > 0:
                        try:
                            len1 = record1.var_col_off[var_list] - record1.var_col_off[var_list - 1]  # 常出错地方
                        except IndexError:
                            continue
                    var_list = var_list + 1
                    col_data_1 = in_data[var_off:var_off + len1]
                    var_off += len1
                col_data2 = data_type.data_type(col_data_1, table.col[ii3])  # 列数据解析

            elif col_is_null == 1:  # 如果列为空
                if table.col[ii3].varlen_is == 0:  # 是定长
                    if table.col[ii3].col_x_type in (3, 239):
                        len1 = len1 * 1
                    col_off += len1
                elif table.col[ii3].varlen_is == 1:  # 是变长
                    var_list = var_list + 1
                    len1 = 0
                col_data2 = ''  # 列数据解析

            record1.col_data1.append(col_data2)

        page1.record.append(record1)  # 把记录放到 页面的记录容器里,会很多
    return page1


# 删除记录解析， bits, 行溢出 还没加(重构前有)
def record_2(f, in_data, table):  # 解析 compact 记录  in_data(输入页面数据),table(表结构)
    page1 = table_struct.st_page()
    chk_type = s("B", in_data[5:6])  # 报这个错的，一般是 此页面数据没找到
    if chk_type[0] % 4 == 1:  # 处理残缺页校验的页数据
        in_data = page_handle(in_data)
    data1 = s("<4B2HIHHIHHI2HIH", in_data[0:38])
    page1.page_no = data1[15]
    page1.page_type = data1[1]
    page1.obj_id = data1[12]
    page1.ind_id = data1[5]
    len_0 = s("<H", in_data[6:8])
    off_set = 96
    while off_set < 8190:  # 在页面中扫描所有 记录头有删除标志的记录的位置
        data = in_data[off_set:8190]
        if len_0[0] == 256:
            off = data.find(b'\x30\x00', 0, -1)  # 30, 3c,70
        else:
            off = data.find(b'\x30\x00', 0, -1)
        if off != -1:
            off = off + off_set
            len_2 = s("<H", in_data[off + 2:off + 4])
            off_1 = off + len_2[0]
            if off_1 > 8190:
                off_set = off + 1
                continue
            len_2 = s("<H", in_data[off_1:off_1 + 2])  # 记录中的列数
            if len_2[0] == table.col_sum:
                page1.page_slot.append(off)
                off_set = off_1
            else:
                off_set = off + 1
        else:
            off_set = off_set + 1

    for i in range(len(page1.page_slot)):  # 页中的所有记录
        record1 = table_struct.st_record()
        off_set = page1.page_slot[i]  # 记录偏移
        try:
            header1 = s("<2h", in_data[off_set:off_set + 4])  # 记录头
        except struct.error:
            continue
        header_a = header1[0]  # 记录类型
        header_b = header1[1]
        # if header_b != 356:
        #     continue
        # if header_a in (12,28,60,167):         # 删除记录
        #     continue
        off1 = header1[1] + off_set  # 列信息的位置偏移
        try:
            col_sum = s('<H', in_data[off1:off1 + 2])
        except struct.error:
            continue
        record1.col_sum = col_sum[0]  # 此记录的总列数
        null_map_len = (record1.col_sum + 7) // 8  # null位图的长度
        null_map_len1 = (table.col_sum + 7) // 8
        if null_map_len > 10:
            continue
        frm = '<' + str(null_map_len) + 'B'
        null_map = s(frm, in_data[off1 + 2:off1 + 2 + null_map_len])  # null位图  +++++++++++++++++++++++++++++++++
        null_map1 = []
        for i1 in range(null_map_len1):
            for i2 in range(0, 8):
                if i1 * 8 + i2 < record1.col_sum:
                    a = null_map[i1] >> i2
                    col_is_null = a % 2  # 此列是否为 null，1=null
                elif i1 * 8 + i2 >= record1.col_sum:
                    col_is_null = 1
                null_map1.append(col_is_null)  # 记录中的位图是描述的原表结构的null情况
        record1.null_map = null_map1  # null位图  ++++++++++++++++++++++++++++++++

        var_col_sum = s('<H', in_data[off1 + 2 + null_map_len:off1 + 4 + null_map_len])  # 变长列数量
        if table.var_len_sum == 0 or header_a == 16:  # 无变长列
            var_col_sum = [0, ]
        record1.var_col_sum = var_col_sum[0]  # 变长列数量
        for ii in range(record1.var_col_sum):  # 变长列 偏移
            try:
                var_col_off1 = s('<H', in_data[off1 + 4 + null_map_len + 2 * ii:off1 + 6 + null_map_len + 2 * ii])
            except struct.error:
                continue
            var_col_off = var_col_off1[0]
            var_col_over = 0
            if var_col_off > 32768:  # 字段 溢出,标志
                var_col_off -= 32768
                var_col_over = 1
            record1.var_col_over.append(var_col_over)
            record1.var_col_off.append(var_col_off)

        # if page1.obj_id == 1 :                      # 测试用
        #     print('i:%d, page_no:%d,obj_id:%d,ind_id:%d,level:%d,rec_sum:%d, off_set:%d,col_sum:%d,null_map_len:%d,null_map_len1:%d'%\
        #     (i,page1.page_no,page1.obj_id,page1.ind_id,page1.page_level,page1.rec_sum,off_set,record1.col_sum,null_map_len,null_map_len1))

        # 解析记录的各列数据, 行溢出先略,  ======================================
        col_off = off_set + 4  # 各列的起始偏移
        var_list = 0
        bit_1 = 0;
        bit_2 = 0;
        data_out = 0
        var_off = off1 + null_map_len + 4 + record1.var_col_sum * 2
        for ii3 in range(len(table.col)):
            len1 = table.col[ii3].col_len  # 列数据类型长度
            col_is_null = record1.null_map[ii3]  # 列是否为空
            if col_is_null == 0:  # 如果列 不为空
                if table.col[ii3].varlen_is == 0:  # 如果列 是定长
                    if table.col[ii3].col_x_type in (3, 239):  # nchar, ntext
                        len1 = len1 * 1
                    col_data_1 = in_data[col_off:col_off + len1]
                    col_off += len1
                elif table.col[ii3].varlen_is == 1:  # 是变长
                    if var_list == 0:  # 第一个变长
                        try:
                            len1 = record1.var_col_off[var_list] - header1[1] - (
                                    null_map_len + 4 + record1.var_col_sum * 2)
                        except IndexError:
                            continue
                    elif var_list > 0:
                        try:
                            len1 = record1.var_col_off[var_list] - record1.var_col_off[var_list - 1]  # 常出错地方
                        except IndexError:
                            continue
                    var_list = var_list + 1
                    col_data_1 = in_data[var_off:var_off + len1]
                    var_off += len1
                col_data2 = data_type.data_type(col_data_1, table.col[ii3])  # 列数据解析

            elif col_is_null == 1:  # 如果列为空
                if table.col[ii3].varlen_is == 0:  # 是定长
                    if table.col[ii3].col_x_type in (3, 239):
                        len1 = len1 * 1
                    col_off += len1
                elif table.col[ii3].varlen_is == 1:  # 是变长
                    var_list = var_list + 1
                    len1 = 0
                col_data2 = ''  # 列数据解析

            record1.col_data1.append(col_data2)

        page1.record.append(record1)  # 把记录放到 页面的记录容器里,会很多
    return page1


# 列数据溢出
def col_over(f, data, col_1):
    data_out = ''
    if col_1.col_x_type in (34, 99, 35):  # text,ntext,img
        rowid = s("<IHH", data[8:16])
        file_no = rowid[1]
        page_no = rowid[0]
        slot_no = rowid[2]
        #      print(page_no,slot_no)
        f.seek(page_no * 8192)
        in_data = f.read(8192)
        page1, in_data = page_init(in_data)
        #   print(page1.page_no,page1.page_slot)
        if page1.page_type != 3:
            return data_out
        off = page1.page_slot[slot_no]
        rec_over_h = s("<HHQH", in_data[off:off + 14])
        rec_len = rec_over_h[1]
        rec_type = rec_over_h[3]
        if rec_type == 3:  # DATA
            data = in_data[off + 14:off + rec_len]
            try:
                if col_1.col_x_type == 35:
                    data_out = str(data, encoding="gbk").rstrip()
                else:
                    data_out = str(data, encoding="utf-16").rstrip()
            except UnicodeDecodeError:
                data_out = ''
        elif rec_type == 0:  # small
            rec_len_1 = s("<H", in_data[off + 14:off + 16])
            data = in_data[off + 20:off + 20 + rec_len_1[0]]
            try:
                if col_1.col_x_type == 35:
                    data_out = str(data, encoding="gbk").rstrip()
                else:
                    data_out = str(data, encoding="utf-16").rstrip()
            except UnicodeDecodeError:
                data_out = ''
        elif rec_type == 5:  # large
            data_1 = ''
            rec_h_1 = s("<HH", in_data[off + 14:off + 18])
            MaxLinks = rec_h_1[0]  # 最大链接数
            CurLinks = rec_h_1[1]  # 实际链接数
            for i in range(CurLinks):
                rec_h_2 = in_data[off + 20 + i * 12:off + 36 + i * 12]
                data_out_1 = col_over(f, rec_h_2, col_1)
                #    print(data_out_1)
                data_1 = data_1 + data_out_1
            data_out = data_1
    elif col_1.col_x_type in (165, 231, 167):  # varchar,nvarchar,varbinary
        data_1 = ''
        len_1 = len(data) // 12
        for i in range(1, len_1):
            rec_h_2 = data[(i * 12 - 4):(i + 1) * 12]
            rowid = s("<IHH", rec_h_2[8:16])
            id = col_1.col_x_type
            if col_1.col_x_type == 167:
                col_1.col_x_type = 35
            else:
                col_1.col_x_type = 99
            data_out_1 = col_over(f, rec_h_2, col_1)
            col_1.col_x_type = id
            data_1 = data_1 + data_out_1
        #    print(data_1)
        data_out = data_1

        print('varchar 行溢出，')

    return data_out
